pageserver/src/tenant/metadata.rs

```diff
@@ -399,126 +431,116 @@ mod tests {
+        assert_eq!(
+            deserialized_metadata.body, expected_metadata.body,
+            "Metadata of the old version {} should be upgraded to the latest version {}",
+            METADATA_OLD_FORMAT_VERSION_V2, METADATA_FORMAT_VERSION
+        );
+    }
 
-        let expected_bincode_bytes = {
-            let mut temp = vec![];
-            let len_bytes = metadata_bytes.len().to_be_bytes();
-            temp.extend_from_slice(&len_bytes);
-            temp.extend_from_slice(&metadata_bytes);
-            temp
-        };
-        assert_eq!(metadata_bincode_be_bytes, expected_bincode_bytes);
-
-        let deserialized_metadata = TimelineMetadata::des(&metadata_bincode_be_bytes).unwrap();
-        // Deserialized metadata has the metadata header, which is different from the serialized one.
-        //   Reference: TimelineMetaData::to_bytes()
-        let expected_metadata = {
-            let mut temp_metadata = original_metadata;
-            let body_bytes = temp_metadata
-                .body
-                .ser()
-                .expect("Cannot serialize the metadata body");
-            let metadata_size = METADATA_HDR_SIZE + body_bytes.len();
-            let hdr = TimelineMetadataHeader {
-                size: metadata_size as u16,
+    #[test]
+    fn test_roundtrip_metadata_v3() {
+        let metadata_v3 = TimelineMetadata {
+            hdr: TimelineMetadataHeader {
+                checksum: 0,
+                size: 0,
                 format_version: METADATA_FORMAT_VERSION,
-                checksum: crc32c::crc32c(&body_bytes),
-            };
-            temp_metadata.hdr = hdr;
-            temp_metadata
+            },
+            body: TimelineMetadataBodyV3 {
+                disk_consistent_lsn: Lsn(0x200),
+                prev_record_lsn: Some(Lsn(0x100)),
+                ancestor_timeline: Some(TIMELINE_ID),
+                ancestor_lsn: Lsn(0),
+                latest_gc_cutoff_lsn: Lsn(0),
+                initdb_lsn: Lsn(0),
+                pg_version: 16,
+            },
         };
-        assert_eq!(deserialized_metadata, expected_metadata);
-    }
 
-    #[test]
-    fn test_metadata_bincode_serde_ensure_roundtrip() {
-        let original_metadata = TimelineMetadata::new(
+        let metadata_bytes = metadata_v3
+            .to_bytes()
+            .expect("Should serialize correct metadata to bytes");
+
+        // This should deserialize to the latest version format
+        let deserialized_metadata = TimelineMetadata::from_bytes(&metadata_bytes)
+            .expect("Should deserialize its own bytes");
+
+        let expected_metadata = TimelineMetadata::new(
             Lsn(0x200),
             Some(Lsn(0x100)),
             Some(TIMELINE_ID),
             Lsn(0),
             Lsn(0),
             Lsn(0),
-            // Any version will do here, so use the default
-            crate::DEFAULT_PG_VERSION,
+            16,
         );
-        let expected_bytes = vec![
```

Comment:
Can you put a json literal here or something else? We should not accidentially break the format and there is few tests for that in the testsuite, so I'd like to avoid reducing the number.

Reply:
sounds good, adding a regression test...

Reply:
added https://github.com/neondatabase/neon/pull/7663/commits/21f3984fccbe377931f02bdfa635d5564f7eed58